content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,79 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- Tests should check not only valid inputs, but **invalid inputs and edge conditions should be tested too** (if problem allows them). `null` inputs, empty arrays, empty strings, etc. all should be tested with fixed tests, unless kata description explicitly says that such inputs won't be provided and don't have to be handled by a solution. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are not common testing practice in "real life" coding, and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass.
+
+- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by a chance.
+- **Randomly generated inputs should be different every time the test suite is run** (with some exceptions, see below). Be sure to seed your RNGs appropriately, and use your random functions correctly, so values used to test user solution are unpredictable. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- **Order of random tests should be non-deterministic.** Not only inputs should be randomized, but their order too. It should not be the case that, for example, 10 valid, random inputs is tested as the first step, and after that, tests generate and verify 10 invalid inputs. Order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios, and shuffle them.
+- Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use predefined, hardcoded set of inputs (and possibly expected outputs too), and before tests are run, shuffle them. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for the result of the test.
```

Comment:
```suggestion
- Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too), and shuffle them before the tests are run. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
```